app.js

```diff
@@ -108,43 +120,67 @@ router.post('/account/:phoneNumber/:accountId/requestCode', async ctx => {
-        return nacl.sign.detached.verify(hash, Buffer.from(signature, 'base64'), bs58.decode(publicKey));
-    });
 };
 
-// TODO: Different endpoints for setup and recovery
-router.post('/account/:phoneNumber/:accountId/validateCode', async ctx => {
-    const { phoneNumber, accountId } = ctx.params;
-    const { securityCode, signature, publicKey } = ctx.request.body;
+function recoveryMethodsFor(account) {
+    return {
+        email: account.email || null,
+        emailAddedAt: account.emailAddedAt || null,
+        phoneAddedAt: account.phoneAddedAt || null,
+        phoneNumber: account.phoneNumber || null,
+        phraseAddedAt: account.phraseAddedAt || null,
+    };
+}
+
+router.post('/account/recoveryMethods', checkAccountOwnership, async ctx => {
+    const { accountId } = ctx.request.body;
+    const account = await models.Account.findOne({ where: { accountId } });
+    ctx.body = recoveryMethodsFor(account || {});
+});
 
-    const account = await models.Account.findOne({ where: { accountId, phoneNumber } });
-    if (!account || !account.securityCode || account.securityCode != securityCode) {
-        ctx.throw(401);
+const recoveryMethodUpdaters = {
+    phone: account => account.update({ phoneNumber: null, phoneAddedAt: null }),
+    email: account => account.update({ email: null, emailAddedAt: null }),
+    phrase: account => account.update({ phraseAddedAt: null })
+};
+
+async function checkRecoveryMethod(ctx, next) {
+    const { recoveryMethod } = ctx.request.body;
+    const allRecoveryMethods = Object.keys(recoveryMethodUpdaters);
+    if (allRecoveryMethods.includes(recoveryMethod)) {
+        await next();
+        return;
     }
-    if (!account.confirmed) {
-        const nearAccount = await ctx.near.account(accountId);
-        const isSignatureValid = await verifySignature(nearAccount, securityCode, signature);
-        if (!isSignatureValid) {
-            ctx.throw(401);
-        }
-        await account.update({ securityCode: null, confirmed: true });
-    } else {
-        await (await ctx.near.account(accountId)).addKey(publicKey);
-        await account.update({ securityCode: null });
+    ctx.throw(400, `Given recoveryMethod '${recoveryMethod}' invalid; must be one of: ${allRecoveryMethods.join(', ')}`);
+}
+
+async function checkAccountInDB(ctx, next) {
+    const { accountId } = ctx.request.body;
+    const account = await models.Account.findOne({ where: { accountId } });
+    if (account) {
+        await next();
+        return;
     }
+    ctx.throw(404, `Could not find account with accountId: '${accountId}'`);
+}
 
-    ctx.body = {};
+router.post('/account/deleteRecoveryMethod', checkRecoveryMethod, checkAccountInDB, checkAccountOwnership, async ctx => {
+    const { accountId, recoveryMethod } = ctx.request.body;
+    const account = await models.Account.findOne({ where: { accountId } });
+    await recoveryMethodUpdaters[recoveryMethod](account);
```

Comment:
It's not a common pattern (only used once) so I feel like this indirection is confusing. Maybe just use `switch`?

Reply:
also somewhat confusing that these are "updaters", but actually they all delete info

Reply:
if you want to make it generic – it makes sense to define separate `RecoveryMethod` or `Key` entity and then keep a list of these in `Account`

Reply:
I've renamed it to `recoveryMethodDeleters`, but I want to keep this pattern rather than using a switch statement. Here's what a switch statement would look like:

```js
const recoveryMethods = ['phone', 'email', 'phrase'];
async function checkRecoveryMethod(ctx, next) {
    const { recoveryMethod } = ctx.request.body;
    if (recoveryMethods.includes(recoveryMethod)) {
        await next();
        return;
    }
    ctx.throw(400, `Given recoveryMethod '${recoveryMethod}' invalid; must be one of: ${recoveryMethods.join(', ')}`);
}
router.post(
    '/account/deleteRecoveryMethod',
    checkRecoveryMethod,
    withAccount,
    checkAccountOwnership,
    async ctx => {
        switch (ctx.request.body.recoveryMethod) {
        case 'phone':
            await ctx.account.update({ phoneNumber: null, phoneAddedAt: null });
            break;
        case 'email':
            await ctx.account.update({ email: null, emailAddedAt: null });
            break;
        case 'phrase':
            await ctx.account.update({ phraseAddedAt: null });
        }
        ctx.body = recoveryMethodsFor(ctx.account);
    }
);
```

While more common, I don't like that this introduces some easy-to-miss coupling between `checkRecoveryMethod` and the body of the route – if we add a new recovery method, we will need to add an entry to `recoveryMethods` and adding a branch to the `switch` statement.

I sort of like the idea of adding this logic to the `Account` model, but right now the logic is small and contained right next to the place it's needed. Adding it to `Account` would clutter our global `Account` model and would, imo, result in harder-to-understand code than either option discussed so far.